[ConstraintSystem] Extend the ExplicitlySpecifyGenericArguments fix to cover all cases of missing generic parameters.

hborla opened the pull request:
* Default holes to `Unresolved`, and move this defaulting inside of `getPotentialBindings`. 
  * Add an extra flag to `PotentialBindings::BindingScore` in order to never pick the bindings for a hole over any other bindings. Similarly, we never choose the bindings for a hole over a disjunction in `ComponentStep::take`.
  * This removes the need to add a defaultable constraint inside of `recordHole`.
  * This removes the need for manual hole propagation. With this new strategy, hole propagation happens automatically.
* Allow fixes with the same locator to be coalesced before application rather than throwing away subsequent fixes when recording.
  * This allows us to coalesce missing generic parameters for the same locator into one `ExplicitlySpecifyGenericArguments` fix, even though the individual parameters will be recorded as missing at different times.
* Rename `ConstraintSystem::recordHole/isHole` to `recordPotentialHole` and `isPotentialHole`, since we don't know for sure whether a type variable is a hole until it's bound to `unresolved`
* Remove `diagnoseAmbiguousGenericParameters` from CSDiag.

rdar://problem/46427366
rdar://problem/56560500
[SR-9351](https://bugs.swift.org/browse/SR-9351)

hborla committed:
[ConstraintSystem] Extend the ExplicitlySpecifyGenericArguments fix to cover
all cases of missing generic parameters.

In `ComponentStep::take` when there are no bindings or disjunctions, use hole
propagation to default remaining free type variables that aren't for generic
parameters and continue solving. Rather than using a defaultable constraint for
holes, assign a fixed type directly when we have no bindings to try.

hborla committed:
[ConstraintSystem] Allow generic parameters and holes to default to
`Any` in `getPotentialBindings` rather than `ComponentStep::take`.

hborla committed:
[ConstraintSystem] Never choose the potential binding for a hole
in the constraint system over a different binding or disjunction.

In other words, we will only choose to bind a hole to `Any` if there
are no other bindings and no disjunctions.

hborla committed:
[ConstraintSystem] Only apply the ExplicitlySpecifyGenericArguments fix
for generic parameters that are holes.

hborla committed:
[Diagnostics] Allow fixes with the same locator to be coalesced before
application rather than throwing away subsequent fixes.

The "primary fix" decides whether or not to coalesce fixes that have
the same locator.

hborla committed:
[ConstraintSystem] Move hole propagation to `addTypeVariableConstraintsToWorkList`
where we are already gathering constraints.

xedin reviewed (commented):
I'll try to take a closer look tonight but so far it looks great! One idea though - how about we add `TVO_Hole` flag to a type variable so it's easier to identify and handle in potential bindings and other places?

hborla committed:
[Test] Update nonnominal_init.swift with better diagnostics from
hole propagation.

hborla committed:
[ConstraintSystem] Bind holes to UnresolvedType instead of Any.

hborla committed:
[ConstraintSystem] With the new approach for holes, hole propagation happens
automatically.

This commit also renames `ConstraintSystem::recordHole/isHole` to
`recordPotentialHole` and `isPotentialHole` to make it clear that
we don't know for sure whether a type variable is a hole until it's
bound to unresolved.

hborla committed:
[ConstraintSystem] Fix a few places where we should consider
`matchTypes` a success without applying new constraint fixes
due to the presence of holes.

hborla commented:
@swift-ci please smoke test

hborla commented on lib/Sema/CSBindings.cpp:
```diff
@@ -979,31 +989,18 @@ bool TypeVariableBinding::attempt(ConstraintSystem &cs) const {
+
+    if (locator->isForGenericParameter() && type->isHole()) {
+      // Drop `generic parameter '...'` part of the locator to group all of the
+      // missing generic parameters related to the same path together.
```
Note: I'm working on a follow-up refactoring where we do not remove the `generic parameter '...'` part of the locator when creating the `ExplicitlySpecifyGenericArguments` constraint fix. This causes conflicts with other fixes applied for that truncated locator, which causes us to not emit the `generic parameter 'T' could not be inferred` error in certain cases.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -4732,7 +4732,8 @@ bool MissingGenericArgumentsFailure::diagnoseParameter(
       llvm::count_if(cs.DefaultedConstraints,
                      [&GP](const ConstraintLocator *locator) {
-                       return locator->getGenericParameter() == GP;
+                       return (locator->isForGenericParameter() &&
```
Is it necessary to check for `isForGenericParameter` first because `getGenericParameter` would return `nullptr` in case it's not which wouldn't be equal to a given `GP`?

hborla replied:
Here's the implementation of `ConstraintLocator::getGenericParameter`:
```
GenericTypeParamType *ConstraintLocator::getGenericParameter() const {
  return castLastElementTo<LocatorPathElt::GenericParameter>().getType();
}
```
So, if the last element is not `LocatorPathElt::GenericParameter`, this will crash. Of course, I could change the implementation to behave like `TypeVariableType::Implementation::getGenericParameter`, which does do the check first. What do you think?

xedin replied:
I didn't realize that it got re-implemented that way! :) I think it would make sense to check inside and return `nullptr` if type variable is not associated with generic parameter...

xedin replied:
I'm pretty sure there is still a bunch of code that replies on the old behavior too...

xedin commented on lib/Sema/CSFix.cpp:
```diff
@@ -714,6 +714,24 @@ bool ExplicitlySpecifyGenericArguments::diagnose(Expr *root,
+  auto params = getParameters();
+  llvm::SmallVector<GenericTypeParamType *, 4> missingParams{params.begin(),
+                                                             params.end()};
+  for (auto *otherFix : fixes)
```
Nit: could you please add parens for this `for`?

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -98,7 +98,7 @@ class FailureDiagnostic {
   Type resolveType(Type rawType, bool reconstituteSugar = false,
                    bool wantRValue = true) const {
-    auto resolvedType = CS.simplifyType(rawType);
+    auto resolvedType = CS.simplifyType(rawType, /*forDiagnostics*/true);
```
Since all the logic involved in diagnostics goes through `resolveType` I think it might makes more sense just "re-implement" `simplifyType` here the way we want to keep constraint system in the dark about how we want "holes" printed. WDYT?

hborla replied:
Sure, I'm happy to do that. It did feel weird to put something specific to diagnostics into `simplifyType`, but I needed the recursive behavior. I'll make that change.

xedin reviewed (commented):
Looks great! I have left a couple comments inline.

xedin commented on lib/Sema/CSStep.cpp:
```diff
@@ -329,7 +329,8 @@ StepResult ComponentStep::take(bool prevFailed) {
   auto bestBindings = CS.determineBestBindings();
 
-  if (bestBindings && (!disjunction || (!bestBindings->InvolvesTypeVariables &&
+  if (bestBindings && (!disjunction || (!bestBindings->IsHole &&
```
I thought that idea was to continue attempting type variables until there are no more bindings to attempt. If we disallow bindings which represent holes I don't think we'll be able to solve situations like this:

```swift
struct S {
  func foo<T>(_ a: T, _ b: Int) -> Void {}
  func foo(_ a: String) -> Void {}

  func test() -> Void {
    foo(.hello, 42)
  }
}
```

Where `T` of `foo` is going to be a base type for `.hello` which means that we probably have to attempt `T := UnresolvedType` to be able to propagate holes to the result type variable of the `unresolved member` reference and the rest of the constraint system.

hborla replied:
This idea is to not choose a hole binding if there is a disjunction. So, we will always pick an available disjunction over a binding for a hole. If there is no disjunction, we will attempt the hole binding. The logic is kind of hard to read though :/ 

Your test case diagnoses:
```
/Users/hborla/Desktop/test.swift:69:5: error: generic parameter 'T' could not be inferred
    foo(.hello, 42)
    ^
/Users/hborla/Desktop/test.swift:65:8: note: in call to function 'foo'
  func foo<T>(_ a: T, _ b: Int) -> Void {}
       ^
```

hborla replied:
In other words, we only choose a hole binding if there are no other bindings to try _and_ no disjunctions to try

xedin replied:
Ah right, sorry I misread `!disjunction` bit.

I gets correctly diagnosed even right now but it's because of CSDiag trying re-typecheck it piecemeal. What I'm wondering is - with your changes does it get diagnosed by a fix without any CSDiag involvement?

hborla replied:
Sorry, yes, I meant that I just tested it with my change and it is diagnosed properly without going through CSDiag.

xedin replied:
Great! Thank you for conforming!

hborla committed:
[ConstraintLocator] Change `ConstraintLocator::getGenericParameter` to return
`nullptr` if the locator path does not terminate at a `GenericParameter` element.

hborla committed:
[Diagnostics] Change `FailureDiagnostic::resolveType` to replace
holes with generic parameters directly when possible.

hborla commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -95,10 +95,27 @@ class FailureDiagnostic {
   Type getType(Expr *expr, bool wantRValue = true) const;
   Type getType(const TypeLoc &loc, bool wantRValue = true) const;
 
+  Type simplifyType(Type type) const  {
```
Let's not add `simplifyType` to a `FailureDiagnostic`, let's just inline the call to `simplifyTypeImpl` there. It would be easier there was only one way to "resolve" a type.

xedin commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -3024,6 +3024,11 @@ class ConstraintSystem {
 
+  /// Simplifies a type by replacing type variables with the result of
+  /// \c getFixedTypeFn and performing lookup on dependent member types.
+  Type simplifyTypeImpl(Type type,
```
This method should probably be `private:`

hborla replied:
`Solution` calls this function. Should `Solution` be a friend of `ConstraintSystem`?

xedin replied:
That would make sense.

hborla commented:
@swift-ci please smoke test

xedin reviewed (approved):
Looks great!

hborla commented:
@swift-ci please smoke test

DougGregor reviewed (approved):
This looks great, thank you!

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test
